smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,390 @@
+ */
+
+package software.amazon.smithy.codegen.core;
+
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.SourceLocation;
+import software.amazon.smithy.model.node.ArrayNode;
+import software.amazon.smithy.model.node.ExpectationNotMetException;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.StringNode;
+import software.amazon.smithy.model.node.ToNode;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.utils.MapUtils;
+import software.amazon.smithy.utils.SmithyBuilder;
+import software.amazon.smithy.utils.ToSmithyBuilder;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.io.Writer;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Class that represents the contents of a Smithy trace file.
+ * TraceFile's require a smithyTrace file version number, {@link ArtifactMetadata}, and
+ * {@link Map} from {@link ShapeId} to a List of {@link ShapeLink} objects. TraceFile's
+ * optionally have a {@link ArtifactDefinitions} object.
+ * <p>
+ * TraceFile handles parsing, serialization and deserialization of a Smithy trace file.
+ * </p>
+ * <p>
+ * The smithyTrace {@link String} contains the Smithy trace file version number.
+ * </p>
+ * <p>
+ * The shapes {@link Map} provides a mapping of absolute Smithy shape IDs to a list
+ * of shape link objects. A single Smithy shape can be responsible for generating
+ * multiple components in the target artifact.
+ * </p>
+ */
+public final class TraceFile implements ToNode, ToSmithyBuilder<TraceFile> {
+    public static final String SMITHY_TRACE_TEXT = "smithyTrace";
+    public static final String ARTIFACT_TEXT = "artifact";
+    public static final String DEFINITIONS_TEXT = "definitions";
+    public static final String SHAPES_TEXT = "shapes";
+    public static final String SMITHY_TRACE_VERSION = "1.0";
+
+    private String smithyTrace;
+    private ArtifactMetadata artifactMetadata;
+    private ArtifactDefinitions artifactDefinitions; //Optional
+    private Map<ShapeId, List<ShapeLink>> shapes;
+    private SourceLocation sl = new SourceLocation("");
+
+    private TraceFile(Builder builder) {
+        smithyTrace = SmithyBuilder.requiredState(SMITHY_TRACE_TEXT, builder.smithyTrace);
+        artifactMetadata = SmithyBuilder.requiredState(ARTIFACT_TEXT, builder.artifactMetadata);
+        shapes = SmithyBuilder.requiredState(SHAPES_TEXT, MapUtils.copyOf(builder.shapes));
```

Comment:
I don't think you need this requirement check here. The `shapes` map is defaulted to an empty map and [`SmithyBuilder.requiredState` works by checking for `null`](https://github.com/awslabs/smithy/blob/master/smithy-utils/src/main/java/software/amazon/smithy/utils/SmithyBuilder.java#L43), meaning this ends up wasted.

The file must contain an entry for `shapes`, which [you're already expecting here.](https://github.com/awslabs/smithy/pull/492/files#diff-340d4054b939dec6834ee291ebe606fbR114)